scripts/DialogData.py

```diff
@@ -14,44 +14,193 @@
-        return self._intents[intentName]
+    def getAllEntities(self):
+        return self._entities
+
+    #   INTENT
+    #******************************************
 
+    def createIntent(self, intent_name):
+        # return intent of a given intentName (if it does not exist it creates an empty one )
+        if intent_name not in self._intents:
+            self._intents[intent_name] = IntentData()
+        return self._intents[intent_name]
 
     def getAllIntents(self):
         return self._intents
 
-
-    def getDomains(self):
+    #   NODE
+    #******************************************
+
+    def createNode(self, node_name, domain_name=None):
+        """ Creates empty node and links it to DialogData.NodeData, returns 0 if already exists
+            - extends _domains if  domainName is not in yet
+        """
+        # Update domain structure, add node to corresponding domain
+        if domain_name is not None:
+            # make sure the domain is registered
+            if domain_name not in self._domains:
+                self._domains[domain_name] = []
+            # make sure the nodeName is remembered within the domain
+            #if node_name not in self._nodes[domain_name]:
+            if node_name not in self._domains[domain_name]:
+                    self._domains[domain_name].append(node_name)
+
+        # Update nodes structure, if node is not there yet - add it
+        if node_name not in self._nodes:
+            self._nodes[node_name] = NodeData()
+            return self._nodes[node_name]
+        else:
+            return None
+
+    def getNode(self, node_name):
+        """  @:returns node of a given name or None
+        """
+        if node_name not in self._nodes:
+            return None
+        return self._nodes[node_name]
+
+    def getAllNodes(self):
+        """  @:returns list of all nodes
+        """
+        return self._nodes
+
+    def updateReferencesNodes(self):
+        """
+            Replaces labels by known target node_names
+            This is implementation of the second path through the dialogData.
+        """
+
+        for node_name in self._nodes: #for each node in the domain
+            nodeData = self.getNode(node_name)
+            if nodeData and nodeData._jumptoTarget:
+                label = nodeData._jumptoTarget
+                if label in self._labelsMap:
+                    node_target= self._labelsMap[label]
+                    nodeData._jumptoTarget= node_target
+                    printf('INFO: Resolving cross reference label:%s -> node_name:%s)\n', label, node_name)
+                else:
+                    printf('INFO: Label:%s not resolve, expecting tahat label is external node_name\n', label)
+
+#   DOMAINS
+#******************************************
+    def getAllDomains(self):
         return self._domains
+
+# createUnique ..
+#******************************************
+    def createUniqueIntentName(self, intent_name):
```

Comment:
This should be merged with `toIntentName` function: https://github.com/IBM/watson-assistant-workbench/blob/7d164be90adae79aebd6fdd0740d9885ad57c3cb/scripts/wawCommons.py#L62

Reply:
createUniqueIntentName is based on toIntentName but adds the checks for existence of the name and auto generation of the modifies to make it unique. Don't we need both of them? Let's discuss further.